File: registration_date.js
const dataPoints = [
    { id: 1, date: new Date(2015, 9, 2) },
    { id: 173870712, date: new Date(2016, 3, 13) },
    { id: 250129497, date: new Date(2017, 5, 3) },
    { id: 401921266, date: new Date(2017, 6, 25) },
    { id: 591702848, date: new Date(2018, 3, 23) },
    { id: 762061130, date: new Date(2019, 6, 27) },
    { id: 989903622, date: new Date(2019, 11, 4) },
    { id: 1000179995, date: new Date(2019, 10, 26) },
    { id: 1100398719, date: new Date(2020, 6, 9) },
    { id: 1338788422, date: new Date(2020, 7, 21) },
    { id: 1366123599, date: new Date(2020, 9, 26) },
    { id: 1400550238, date: new Date(2020, 10, 19) },
    { id: 1500045869, date: new Date(2021, 0, 13) },
    { id: 1600315778, date: new Date(2021, 3, 15) },
    { id: 1700263108, date: new Date(2021, 3, 18) },
    { id: 1809955302, date: new Date(2021, 6, 27) },
    { id: 1904824687, date: new Date(2021, 7, 29) },
    { id: 2004732748, date: new Date(2021, 8, 24) },
    { id: 2115988529, date: new Date(2021, 10, 26) },
    { id: 5004488410, date: new Date(2021, 11, 17) },
    { id: 5367866773, date: new Date(2022, 3, 11) },
    { id: 6040289492, date: new Date(2023, 4, 29) },
    { id: 7275317003, date: new Date(2024, 5, 5) }
];

function calculateAge(estimatedDate) {
    const ageInDays = (new Date() - estimatedDate) / (1000 * 60 * 60 * 24);
    const ageInYears = ageInDays / 365.25;
    console.log(ageInYears);
    roundAge = Math.round(ageInYears);
    if (roundAge < 1) {
        return 1;
    }
    return roundAge;
}

function estimateRegistrationDate(userId, dataPoints) {
    dataPoints.sort((a, b) => a.id - b.id);

    for (let i = 0; i < dataPoints.length - 1; i++) {
        const { id: id1, date: date1 } = dataPoints[i];
        const { id: id2, date: date2 } = dataPoints[i + 1];

        if (id1 <= userId && userId <= id2) {
            const daysBetween = (date2 - date1) / (1000 * 60 * 60 * 24);
            const idsBetween = id2 - id1;
            const growthRate = idsBetween / daysBetween;

            const daysSinceId1 = (userId - id1) / growthRate;
            const estimatedDate = new Date(date1.getTime() + daysSinceId1 * 24 * 60 * 60 * 1000);
            return estimatedDate;
        }
    }
    console.log('User ID not found in data points');
    const { id: idLast, date: dateLast } = dataPoints[dataPoints.length - 1];
    const { id: idSecondLast, date: dateSecondLast } = dataPoints[dataPoints.length - 2];
    const daysBetween = (dateLast - dateSecondLast) / (1000 * 60 * 60 * 24);
    const idsBetween = idLast - idSecondLast;
    const growthRate = idsBetween / daysBetween;

    const daysSinceLast = (userId - idLast) / growthRate;
    const estimatedDate = new Date(dateLast.getTime() + daysSinceLast * 24 * 60 * 60 * 1000);
    return estimatedDate;
}

const userIdToEstimate = 8275317003;
const estimatedDate = estimateRegistrationDate(userIdToEstimate, dataPoints);
const ageInYears = calculateAge(estimatedDate);

console.log(`Estimated registration date for user ID ${userIdToEstimate}: ${estimatedDate.toISOString().split('T')[0]}`);
console.log(`Estimated age of user ID ${userIdToEstimate}: ${ageInYears} years`);

File: registration_date.py
from datetime import datetime, timedelta

# Known data points
data_points = [
    (173870712, datetime(2016, 4, 13)),
    (250129497, datetime(2017, 6, 3)),
    (401921266, datetime(2017, 7, 25)),
    (591702848, datetime(2018, 4, 23)),
    (762061130, datetime(2019, 7, 27)),
    (989903622, datetime(2019, 12, 4)),
    (1000179995, datetime(2019, 11, 26)),
    (1100398719, datetime(2020, 7, 9)),
    (1338788422, datetime(2020, 8, 21)),
    (1366123599, datetime(2020, 10, 26)),
    (1400550238, datetime(2020,11,19)),
    (1500045869, datetime(2021, 1,13)),
    (1600315778, datetime(2021, 4, 15)),
    (1700263108, datetime(2021, 4, 18)),
    (1809955302, datetime(2021, 7, 27)),
    (1904824687, datetime(2021, 8, 29)),
    (2004732748, datetime(2021, 9, 24)),
    (2115988529, datetime(2021, 11, 26)),
    (5004488410, datetime(2021, 12, 17)),
    (5367866773, datetime(2022, 4, 11)),
    (6040289492, datetime(2023, 5, 29)),
    (7275317003, datetime(2024, 6, 5))
]
def calculate_age(estimated_date):
    age_in_days = (datetime.now() - estimated_date).days
    age_in_years = age_in_days / 365.25  # Approximate year length accounting for leap years
    return age_in_years

def estimate_registration_date(user_id, data_points):
    # Sort data points by user_id
    data_points.sort()
    
    # Find the closest data points
    for i in range(len(data_points) - 1):
        id1, date1 = data_points[i]
        id2, date2 = data_points[i + 1]
        
        if id1 <= user_id <= id2:
            # Calculate the growth rate between these points
            days_between = (date2 - date1).days
            ids_between = id2 - id1
            growth_rate = ids_between / days_between
            
            # Estimate the date
            days_since_id1 = (user_id - id1) / growth_rate
            estimated_date = date1 + timedelta(days=days_since_id1)
            return estimated_date

    # If the user_id is beyond the last known data point
    id_last, date_last = data_points[-1]
    growth_rate = (data_points[-1][0] - data_points[-2][0]) / (data_points[-1][1] - data_points[-2][1]).days
    days_since_last = (user_id - id_last) / growth_rate
    estimated_date = date_last + timedelta(days=days_since_last)
    return estimated_date

# User ID to estimate
user_id_to_estimate = 5800091112

# Estimate registration date
estimated_date = estimate_registration_date(user_id_to_estimate, data_points)
print(f"Estimated registration date for user ID {user_id_to_estimate}: {estimated_date.strftime('%Y-%m-%d')}")
print("age", calculate_age(estimated_date))
